File: CRM.Web/src/types/opportunity.ts
import type { Contact } from './contact';
import type { Company } from './company';

// =============================================
// Opportunity Stage Types
// =============================================
export type OpportunityStage =
    | 'Lead'
    | 'Qualified'
    | 'Proposal'
    | 'Negotiation'
    | 'Closed Won'
    | 'Closed Lost'
    | 'Initial' // Backwards compatibility
    | 'Qualification'; // Backwards compatibility

// =============================================
// Main Opportunity Interface
// =============================================
export interface Opportunity {
    id?: number;
    name: string;
    amount: number;
    stage: OpportunityStage;
    probability: number;
    expectedCloseDate: string;

    // Actual Close & Win/Loss
    actualCloseDate?: string;
    wonDate?: string;
    lostDate?: string;
    winReason?: string;
    lostReason?: string;
    winLossNotes?: string;

    // Stage Tracking
    lastStageChangeDate?: string;
    daysInCurrentStage?: number;
    totalDaysToClose?: number;

    // Deal Health & Scoring
    dealScore?: number;
    dealHealth?: 'Hot' | 'Healthy' | 'At Risk' | 'Stalled';
    riskFactors?: string; // JSON array

    // Next Steps
    nextAction?: string;
    nextActionDate?: string;
    nextActionOwner?: string;

    // Competitors
    competitors?: string; // JSON array
    primaryCompetitor?: string;
    competitivePosition?: 'Ahead' | 'Behind' | 'Even' | 'Unknown';

    // Value & Forecasting
    weightedValue?: number;
    recurringValue?: number;
    currency?: string;
    forecastCategory?: 'Pipeline' | 'Best Case' | 'Commit' | 'Closed' | 'Omitted';

    // Relationships
    contactId?: number;
    contact?: Contact;
    companyId?: number;
    company?: Company;
    owner?: string;
    ownerId?: number;

    // Source & Type
    source?: string;
    type?: string;

    // Description & Notes
    description?: string;
    tags?: string; // JSON string array

    // Timestamps
    createdAt?: string;
    lastModifiedAt?: string;

    // Navigation
    stageHistory?: StageHistory[];
    products?: OpportunityProduct[];
}

// =============================================
// Opportunity Product/Line Item
// =============================================
export interface OpportunityProduct {
    id?: number;
    opportunityId: number;
    productName: string;
    productCode?: string;
    quantity: number;
    unitPrice: number;
    discount?: number;
    discountType?: 'Percentage' | 'Fixed';
    totalPrice: number;
    description?: string;
    createdAt?: string;
}

// =============================================
// Stage History
// =============================================
export interface StageHistory {
    id?: number;
    opportunityId: number;
    fromStage: string;
    toStage: string;
    changedAt: string;
    changedByUserId?: number;
    daysInPreviousStage?: number;
}

// =============================================
// Pipeline Card (for Kanban view)
// =============================================
export interface PipelineCard {
    id: number;
    name: string;
    amount: number;
    weightedValue: number;
    probability: number;
    expectedCloseDate: string;
    dealScore?: number;
    dealHealth?: string;
    nextAction?: string;
    owner?: string;
    contactName?: string;
    companyName?: string;
    daysInStage?: number;
    color?: string;
}

// =============================================
// Stats & Analytics
// =============================================
export interface OpportunityStats {
    totalOpportunities: number;
    openOpportunities: number;
    wonOpportunities: number;
    lostOpportunities: number;

    totalPipelineValue: number;
    weightedPipelineValue: number;
    totalWonValue: number;
    totalLostValue: number;

    winRate: number;
    avgDealSize: number;
    avgDaysToClose: number;

    byStage: StageSummary[];
    byForecast: ForecastSummary[];

    atRiskDeals: number;
    hotDeals: number;
}

export interface StageSummary {
    stage: string;
    count: number;
    totalValue: number;
    weightedValue: number;
    color: string;
}

export interface ForecastSummary {
    category: string;
    count: number;
    value: number;
}

export interface VelocityMetrics {
    avgSalesCycle: number;
    dealsClosed30Days: number;
    revenueClosed30Days: number;
    winRate30Days: number;
    avgDaysByStage: { stage: string; avgDaysInStage: number }[];
}

export interface LeaderboardEntry {
    owner: string;
    totalDeals: number;
    wonDeals: number;
    lostDeals: number;
    openDeals: number;
    wonValue: number;
    pipelineValue: number;
    winRate: number;
}

// =============================================
// Constants
// =============================================
export const OPPORTUNITY_STAGES: OpportunityStage[] = [
    'Lead',
    'Qualified',
    'Proposal',
    'Negotiation',
    'Closed Won',
    'Closed Lost'
];

export const STAGE_COLORS: Record<string, string> = {
    'Lead': '#6366F1',
    'Qualified': '#3B82F6',
    'Proposal': '#F59E0B',
    'Negotiation': '#8B5CF6',
    'Closed Won': '#22C55E',
    'Closed Lost': '#EF4444',
    // Backwards compatibility
    'Initial': '#6366F1',
    'Qualification': '#3B82F6'
};

export const DEFAULT_PROBABILITIES: Record<string, number> = {
    'Lead': 10,
    'Qualified': 25,
    'Proposal': 50,
    'Negotiation': 75,
    'Closed Won': 100,
    'Closed Lost': 0,
    'Initial': 10,
    'Qualification': 25
};

export const WIN_REASONS = [
    'Price', 'Features', 'Relationship', 'Brand Trust',
    'Speed', 'Support', 'Customization', 'Integration', 'Other'
];

export const LOSS_REASONS = [
    'Price Too High', 'Lost to Competition', 'No Budget', 'Bad Timing', 'No Decision',
    'Missing Features', 'Poor Fit', 'Champion Left', 'Project Cancelled', 'Other'
];

export const SOURCES = [
    'Website', 'Referral', 'Cold Call', 'Trade Show', 'Partner', 'Social Media',
    'Email Campaign', 'Inbound', 'Outbound', 'Existing Customer', 'Other'
];

export const OPPORTUNITY_TYPES = [
    'New Business', 'Upsell', 'Cross-sell', 'Renewal', 'Expansion', 'Replacement'
];

export const FORECAST_CATEGORIES = [
    'Pipeline', 'Best Case', 'Commit', 'Closed', 'Omitted'
];

export const DEAL_HEALTH_OPTIONS = ['Hot', 'Healthy', 'At Risk', 'Stalled'];

File: CRM.Api/Models/Opportunity.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CRM.Api.Models
{
    public class Opportunity
    {
        public int Id { get; set; }

        [Required]
        [StringLength(200)]
        public string Name { get; set; } = string.Empty;

        [Column(TypeName = "decimal(18,2)")]
        public decimal Amount { get; set; }

        [StringLength(50)]
        public string Stage { get; set; } = "Lead";
        // Stages: Lead, Qualified, Proposal, Negotiation, Closed Won, Closed Lost

        public double Probability { get; set; } = 10; // Percentage (0-100)

        public DateTime ExpectedCloseDate { get; set; }

        // =============================================
        // Actual Close & Win/Loss Tracking
        // =============================================
        public DateTime? ActualCloseDate { get; set; }
        
        public DateTime? WonDate { get; set; }
        public DateTime? LostDate { get; set; }
        
        [StringLength(100)]
        public string? WinReason { get; set; } // Price, Features, Relationship, etc.
        
        [StringLength(100)]
        public string? LostReason { get; set; } // Price, Competition, Timing, No Budget, etc.
        
        [StringLength(500)]
        public string? WinLossNotes { get; set; } // Detailed explanation

        // =============================================
        // Stage & Duration Tracking
        // =============================================
        public DateTime? LastStageChangeDate { get; set; }
        public int DaysInCurrentStage { get; set; } = 0;
        
        // Total days from Lead to Close
        public int TotalDaysToClose { get; set; } = 0;

        // =============================================
        // Deal Health & Scoring
        // =============================================
        [Range(0, 100)]
        public int DealScore { get; set; } = 50; // 0-100 health score
        
        [StringLength(20)]
        public string? DealHealth { get; set; } = "Healthy"; // At Risk, Healthy, Hot
        
        [StringLength(500)]
        public string? RiskFactors { get; set; } // JSON array of risk factors

        // =============================================
        // Next Steps & Actions
        // =============================================
        [StringLength(500)]
        public string? NextAction { get; set; }
        
        public DateTime? NextActionDate { get; set; }
        
        [StringLength(200)]
        public string? NextActionOwner { get; set; }

        // =============================================
        // Competitors
        // =============================================
        [StringLength(1000)]
        public string? Competitors { get; set; } // JSON array [{name, strength, weakness}]
        
        [StringLength(200)]
        public string? PrimaryCompetitor { get; set; }
        
        [StringLength(50)]
        public string? CompetitivePosition { get; set; } // Ahead, Behind, Even, Unknown

        // =============================================
        // Value & Forecasting
        // =============================================
        [Column(TypeName = "decimal(18,2)")]
        public decimal WeightedValue { get; set; }
        
        [Column(TypeName = "decimal(18,2)")]
        public decimal? RecurringValue { get; set; } // Monthly/Annual recurring revenue
        
        [StringLength(20)]
        public string? Currency { get; set; } = "USD";
        
        // Forecast category
        [StringLength(50)]
        public string? ForecastCategory { get; set; } // Pipeline, Best Case, Commit, Omitted

        // =============================================
        // Relationships
        // =============================================
        public int? ContactId { get; set; }
        public Contact? Contact { get; set; }
        
        public int? CompanyId { get; set; }
        public Company? Company { get; set; }
        
        [StringLength(200)]
        public string? Owner { get; set; } // Sales rep name
        
        public int? OwnerId { get; set; } // User ID when auth is implemented

        // =============================================
        // Source & Type
        // =============================================
        [StringLength(100)]
        public string? Source { get; set; } // Website, Referral, Cold Call, etc.
        
        [StringLength(100)]
        public string? Type { get; set; } // New Business, Upsell, Renewal, etc.

        // =============================================
        // Description & Notes
        // =============================================
        [StringLength(2000)]
        public string? Description { get; set; }
        
        [StringLength(1000)]
        public string? Tags { get; set; } // JSON array

        // =============================================
        // Timestamps
        // =============================================
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? LastModifiedAt { get; set; } = DateTime.UtcNow;

        // =============================================
        // Navigation Properties
        // =============================================
        public ICollection<StageHistory>? StageHistory { get; set; }
        public ICollection<OpportunityProduct>? Products { get; set; }

        [NotMapped]
        public IEnumerable<CustomFieldValue>? CustomValues { get; set; }
    }

    // =============================================
    // Opportunity Product/Line Item
    // =============================================
    public class OpportunityProduct
    {
        public int Id { get; set; }
        
        public int OpportunityId { get; set; }
        public Opportunity? Opportunity { get; set; }
        
        [Required]
        [StringLength(200)]
        public string ProductName { get; set; } = string.Empty;
        
        [StringLength(50)]
        public string? ProductCode { get; set; }
        
        public int Quantity { get; set; } = 1;
        
        [Column(TypeName = "decimal(18,2)")]
        public decimal UnitPrice { get; set; }
        
        [Column(TypeName = "decimal(18,2)")]
        public decimal? Discount { get; set; } // Percentage or fixed amount
        
        [StringLength(20)]
        public string? DiscountType { get; set; } = "Percentage"; // Percentage or Fixed
        
        [Column(TypeName = "decimal(18,2)")]
        public decimal TotalPrice { get; set; }
        
        [StringLength(500)]
        public string? Description { get; set; }
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }

    // =============================================
    // Opportunity Stage Constants
    // =============================================
    public static class OpportunityStages
    {
        public const string Lead = "Lead";
        public const string Qualified = "Qualified";
        public const string Proposal = "Proposal";
        public const string Negotiation = "Negotiation";
        public const string ClosedWon = "Closed Won";
        public const string ClosedLost = "Closed Lost";

        public static readonly string[] All = { Lead, Qualified, Proposal, Negotiation, ClosedWon, ClosedLost };

        public static readonly Dictionary<string, int> DefaultProbabilities = new()
        {
            { Lead, 10 },
            { Qualified, 25 },
            { Proposal, 50 },
            { Negotiation, 75 },
            { ClosedWon, 100 },
            { ClosedLost, 0 }
        };

        public static readonly Dictionary<string, string> Colors = new()
        {
            { Lead, "#6366F1" },       // Indigo
            { Qualified, "#3B82F6" },   // Blue
            { Proposal, "#F59E0B" },    // Amber
            { Negotiation, "#8B5CF6" }, // Purple
            { ClosedWon, "#22C55E" },   // Green
            { ClosedLost, "#EF4444" }   // Red
        };
    }

    // =============================================
    // Predefined Options
    // =============================================
    public static class OpportunityOptions
    {
        public static readonly string[] WinReasons = 
        {
            "Price", "Features", "Relationship", "Brand Trust", "Speed", "Support", "Customization", "Integration", "Other"
        };

        public static readonly string[] LossReasons =
        {
            "Price Too High", "Lost to Competition", "No Budget", "Bad Timing", "No Decision",
            "Missing Features", "Poor Fit", "Champion Left", "Project Cancelled", "Other"
        };

        public static readonly string[] Sources =
        {
            "Website", "Referral", "Cold Call", "Trade Show", "Partner", "Social Media", 
            "Email Campaign", "Inbound", "Outbound", "Existing Customer", "Other"
        };

        public static readonly string[] Types =
        {
            "New Business", "Upsell", "Cross-sell", "Renewal", "Expansion", "Replacement"
        };

        public static readonly string[] ForecastCategories =
        {
            "Pipeline", "Best Case", "Commit", "Closed", "Omitted"
        };

        public static readonly string[] DealHealthOptions =
        {
            "Hot", "Healthy", "At Risk", "Stalled"
        };
    }
}
